### PR TITLE
Proper font support for Linux and OS X

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
This introduces proper font support for both Linux and OS X through the use of FontConfig.

FontConfig is pretty standard on Linux, but on OS X it will need to be installed through e.g. Homebrew. For this reason, it may be better to replace it with an Objective-C implementation. I know nothing about Cocoa, though -- maybe someone else would like to pick that up?

As FontConfig makes use of a font's real name instead of a filename, I have amended the `TTFFontDescriptor` struct to contain names as well.

I don't know what the conventions are regarding `cmake`, but I may have put the FontConfig dependency in the wrong place. It works as is, however.

For Windows, this means the existing implementation is pretty much retained as it was -- the relevant bits have just been moved to `windows.c` instead.

### CODE REVIEW

**Comment by IntelOrca** [comment]
On Windows, it should use [SHGetKnownFolderPath](https://msdn.microsoft.com/en-us/library/windows/desktop/bb762188(v=vs.85).aspx) with ```FOLDERID_Fonts```.
You will need to do the necessary wchar to utf8 conversion, see ```platform_resolve_user_data_path```.

File: src/platform/posix.c [hunk 1]
```diff
@@ -757,4 +758,31 @@ uint8 platform_get_locale_temperature_format(){
 	return TEMPERATURE_FORMAT_C;
 }
 
+bool platform_get_font_path(TTFFontDescriptor *font, utf8 *buffer)
+{
```
Comment (LinusU) [inline src/platform/posix.c]: The other functions in this file has the `{` on the same row as the function declaration

**Comment by janisozaur** [comment]
I'll test it soon and try to update required packages.

**Comment by Gymnasiast** [comment]
Works for me.

**Comment by AaronVanGeffen** [comment]
I have added implementations for both Windows and OS X in the last two commits. As a result, OS X now uses OS X's Cocoa APIs to do font detection -- it no longer needs FontConfig.

@janisozaur The CMakeFile needs to be adjusted to remove the dependency for OS X. As you improved it earlier, how would you make it conditional, i.e. only for Linux? (PR welcome if that's easier.)

**Comment by janisozaur** [comment]
@AaronVanGeffen as there is no `LINUX` defined, either check for `CMAKE_SYSTEM_NAME` (see https://github.com/OpenRCT2/OpenRCT2/blob/develop/CMakeLists_mingw.txt#L1 and https://github.com/OpenRCT2/OpenRCT2/blob/develop/CMakeLists_mingw.txt#L16) or `if (UNIX and NOT APPLE)` should be good enough for now, I think, or something of that kind.

**Comment by AaronVanGeffen** [comment]
With 43e193b in, I think this should be pretty much done.

Travis seems happy enough when doing the Linux and Docker jobs, but appears to be failing to build through `mingw`. Apparently, [`SHGetKnownFolderPath`](https://msdn.microsoft.com/en-us/library/windows/desktop/bb762188%28v=vs.85%29.aspx) is undefined, even though the relevant header `shlobj.h` is included at the top of `windows.c`. This API was introduced in Windows Vista, so I'm guessing `mingw` is targetting an older version.

What would be the best way to solve this? We could use [`SHGetFolderPath`](https://msdn.microsoft.com/en-us/library/windows/desktop/bb762181%28v=vs.85%29.aspx) instead, but seeing as that's a deprecated API, I would prefer not to.

**Comment by duncanspumpkin** [comment]
@AaronVanGeffen i maybe wrong but aren't the mingw builds the only ones that handle windows xp. I wouldn't want to loose that compatibility for something like this. I would just use the old SHGetFolderPath and leave a note explaining why.

**Comment by IntelOrca** [comment]
@AaronVanGeffen I think you should just leave it as it is. The mingw builds have only exited so that developers on linux and osx could compile the game and run it under wine. Now that it can be built natively for both of those platforms, I don't think anyone is still using mingw.

**Comment by janisozaur** [comment]
I've had a look at possible solution to this with mingw and sadly, I couldn't find any.

We already have a small hack for mingw (SDL) bug somewhere with window handling code, how about doing a mingw version here? Either `SHGetFolderPath` or simply returning `C:\Windows\Fonts\` with a `log_warning`.

Otherwise we'll have to ditch mingw builds, which would be a loss, albeit a small one with native builds now available for all three major platforms.

**Comment by AaronVanGeffen** [comment]
The hack sounds like a good compromise -- I agree it would be a shame to lose MINGW compatibility over something like this. eda7331 introduces one that can easily be removed if/when MINGW introduces the API in their compiler. Looks like Travis is happy now, too.

On an unrelated note: Merry Christmas, everyone.

**Comment by janisozaur** [comment]
Once https://github.com/AaronVanGeffen/OpenRCT2/pull/2 is merged, I'm okay with this change.

Good job @AaronVanGeffen, quite impressive considering you haven't had chance to test it on Linux!

**Comment by janisozaur** [comment]
An important note for Linux side of things (in case of new issues):

Arial and etc. are proprietary and don't come preloaded with Linux. We have libre versions of these fonts, but they, quite obviously, are not pixel-perfect with originals. User can usually install these fonts as a package or provide them manually or continue with libre versions of these. Each distribution ma have its own substitutions, which may cause text to be misaligned from what you see on Windows.

I'll try to update wiki with some information on that once this gets merged.

**Comment by AaronVanGeffen** [comment]
Thanks @janisozaur. I appreciate  you taking the time to look into the patch.

As for compatibility, I think Linux has Liberation Sans set as the default alternative for Arial, right? It's intended to be metrically compatible with Arial, so that should be fine.

**Comment by IntelOrca** [comment]
So are well all happy with this now?

**Comment by Gymnasiast** [comment]
I am, in any case.

**Comment by janisozaur** [comment]
image

**Pull request merged**